In [167]:
#Import all relevant libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import zscore

## This statement allows the visuals to render within your Jupyter Notebook.
%matplotlib inline

## Loading the data
We can now load the dataset into pandas using the read_csv() function. This converts the CSV file into a Pandas dataframe.

In [168]:
#Read in the csv file and convert to a Pandas dataframe
world_happiness_2015 = pd.read_csv("2015.csv")
world_happiness_2016 = pd.read_csv("2016.csv")
world_happiness_2017 = pd.read_csv("2017.csv")
world_happiness_2018 = pd.read_csv("2018.csv")
world_happiness_2019 = pd.read_csv("2019.csv")

### Viewing the dataframe
We can get a quick sense of the size of our dataset by using the shape method. This returns a tuple with the number of rows and columns in the dataset.

In [169]:
# Display the first 3 rows of the `world_happiness_2015` dataset
world_happiness_2015.head(3)

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938                        0.48357   

   Generosity  Dystopia Residual  
0     0.29678            2.51738  
1     0.43630            2.70201  
2     0.34139            2.49204

In [170]:
# Display the first 3 rows of the `world_happiness_2016` dataset
world_happiness_2016.head(3)

Country          Region  Happiness Rank  Happiness Score  \
0      Denmark  Western Europe               1            7.526   
1  Switzerland  Western Europe               2            7.509   
2      Iceland  Western Europe               3            7.501   

   Lower Confidence Interval  Upper Confidence Interval  \
0                      7.460                      7.592   
1                      7.428                      7.590   
2                      7.333                      7.669   

   Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
0                   1.44178  1.16374                   0.79504  0.57941   
1                   1.52733  1.14524                   0.86303  0.58557   
2                   1.42666  1.18326                   0.86733  0.56624   

   Trust (Government Corruption)  Generosity  Dystopia Residual  
0                        0.44453     0.36171            2.73939  
1                        0.41203     0.28083            2.69463  
2                        0.14975     0.47678            2.83137

In [171]:
# Display the first 3 rows of the `world_happiness_2017` dataset
world_happiness_2017.head(3)

Country  Happiness.Rank  Happiness.Score  Whisker.high  Whisker.low  \
0   Norway               1            7.537      7.594445     7.479556   
1  Denmark               2            7.522      7.581728     7.462272   
2  Iceland               3            7.504      7.622030     7.385970   

   Economy..GDP.per.Capita.    Family  Health..Life.Expectancy.   Freedom  \
0                  1.616463  1.533524                  0.796667  0.635423   
1                  1.482383  1.551122                  0.792566  0.626007   
2                  1.480633  1.610574                  0.833552  0.627163   

   Generosity  Trust..Government.Corruption.  Dystopia.Residual  
0    0.362012                       0.315964           2.277027  
1    0.355280                       0.400770           2.313707  
2    0.475540                       0.153527           2.322715

In [172]:
# Display the first 3 rows of the `world_happiness_2018` dataset
world_happiness_2018.head(3)

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.632           1.305           1.592   
1             2            Norway  7.594           1.456           1.582   
2             3           Denmark  7.555           1.351           1.590   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.874                         0.681       0.202   
1                    0.861                         0.686       0.286   
2                    0.868                         0.683       0.284   

   Perceptions of corruption  
0                      0.393  
1                      0.340  
2                      0.408

In [173]:
# Display the first 3 rows of the `world_happiness_2019` dataset
world_happiness_2019.head(3)

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.769           1.340           1.587   
1             2           Denmark  7.600           1.383           1.573   
2             3            Norway  7.554           1.488           1.582   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.986                         0.596       0.153   
1                    0.996                         0.592       0.252   
2                    1.028                         0.603       0.271   

   Perceptions of corruption  
0                      0.393  
1                      0.410  
2                      0.341

In [174]:
# Print the number of rows and columns in the each dataset
print("\nNumber of Rows and Columns in: ")
print("- world_happiness_2015 are: ", world_happiness_2015.shape)
print("- world_happiness_2016 are: ", world_happiness_2016.shape)
print("- world_happiness_2017 are: ", world_happiness_2017.shape)
print("- world_happiness_2018 are: ", world_happiness_2018.shape)
print("- world_happiness_2019 are: ", world_happiness_2019.shape)


Number of Rows and Columns in: 
- world_happiness_2015 are:  (158, 12)
- world_happiness_2016 are:  (157, 13)
- world_happiness_2017 are:  (155, 12)
- world_happiness_2018 are:  (156, 9)
- world_happiness_2019 are:  (156, 9)


 We note that:
- The percentage of countries included in the report is about 81.87%, while the percentage of countries not included is about 18.13%. There are about 35 countries not included in the World Happiness Report out of 193 countries due to the lack of sufficient data or difficulties in collecting the necessary information, which makes the report limited to about 158 ​​countries annually.
- The number of participating countries varies slightly over the years.

## 1. Data Profiling:
Data profiling is a comprehensive process of examining the data available in an existing dataset and collecting statistics and information about that data.

In [175]:
# Get the `world_happiness_2015` dataframe column labels
print("All Column Names in world_happiness_2015 DataFrame:\n", world_happiness_2015.columns)

All Column Names in world_happiness_2015 DataFrame:
 Index(['Country', 'Region', 'Happiness Rank', 'Happiness Score',
       'Standard Error', 'Economy (GDP per Capita)', 'Family',
       'Health (Life Expectancy)', 'Freedom', 'Trust (Government Corruption)',
       'Generosity', 'Dystopia Residual'],
      dtype='object')


In [176]:
# Get the `world_happiness_2016` dataframe column labels
print("All Column Names in world_happiness_2016 DataFrame:\n", world_happiness_2016.columns)

All Column Names in world_happiness_2016 DataFrame:
 Index(['Country', 'Region', 'Happiness Rank', 'Happiness Score',
       'Lower Confidence Interval', 'Upper Confidence Interval',
       'Economy (GDP per Capita)', 'Family', 'Health (Life Expectancy)',
       'Freedom', 'Trust (Government Corruption)', 'Generosity',
       'Dystopia Residual'],
      dtype='object')


In [177]:
# Get the `world_happiness_2017` dataframe column labels
print("All Column Names in world_happiness_2017 DataFrame:\n", world_happiness_2017.columns)

All Column Names in world_happiness_2017 DataFrame:
 Index(['Country', 'Happiness.Rank', 'Happiness.Score', 'Whisker.high',
       'Whisker.low', 'Economy..GDP.per.Capita.', 'Family',
       'Health..Life.Expectancy.', 'Freedom', 'Generosity',
       'Trust..Government.Corruption.', 'Dystopia.Residual'],
      dtype='object')


In [178]:
# Get the `world_happiness_2018` dataframe column labels
print("All Column Names in world_happiness_2018 DataFrame:\n", world_happiness_2018.columns)

All Column Names in world_happiness_2018 DataFrame:
 Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption'],
      dtype='object')


In [179]:
# Get the `world_happiness_2019` dataframe column labels
print("All Column Names in world_happiness_2019 DataFrame:\n", world_happiness_2019.columns)

All Column Names in world_happiness_2019 DataFrame:
 Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption'],
      dtype='object')


### There are 3 Main Differences Between Columns Labels Between the Years:

**1- Addition or deletion of columns**
Comparing to 2015

- In 2016, the `Lower Confidence Interval` and `Upper Confidence Interval` columns were added.
- In 2017, the `Region` and Confidence Interval columns were removed, while `Whisker.high` and `Whisker.low` were added. The addition of Whisker may indicate a shift in focus to a more statistical representation, in line with visualizations such as box plots.
- In 2018 and 2019,  `Region`, `Confidence Intervals (Lower and Upper)`, and `Whisker values` were removed.


**2- Renaming of columns:**

In 2017, the column names were changed to a format that uses dots (.) instead of spaces or parentheses. For example:
  - `Health (Life Expectancy)` became `Health..Life.Expectancy`.
  - `Trust (Government Corruption)` became `Trust..Government.Corruption`.
In 2018:
  - `Health (Life Expectancy)` became `Healthy life expectancy`.
  - `Trust (Government Corruption)` became `Perceptions of corruption`.

In 2019:
Same as 2018 changes.

**3. Addition of new statistical columns:**

Such as confidence intervals and upper/lower bounds for happiness.

In 2018:
  - Introduced `Social support` instead of `Family`.
  - Focused on core indicators without additional statistical metrics.

In 2019:
Similar to 2018 with core indicators such as GDP per capita, social support, healthy life expectancy, freedom, generosity, and perceptions of corruption.


In [180]:
# print the information of `world_happiness_2015` dataframe
world_happiness_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 0 to 157
Data columns (total 12 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        158 non-null    object 
 1   Region                         158 non-null    object 
 2   Happiness Rank                 158 non-null    int64  
 3   Happiness Score                158 non-null    float64
 4   Standard Error                 158 non-null    float64
 5   Economy (GDP per Capita)       158 non-null    float64
 6   Family                         158 non-null    float64
 7   Health (Life Expectancy)       158 non-null    float64
 8   Freedom                        158 non-null    float64
 9   Trust (Government Corruption)  158 non-null    float64
 10  Generosity                     158 non-null    float64
 11  Dystopia Residual              158 non-null    float64
dtypes: float64(9), int64(1), object(2)
memory usage: 1

There are no missing values and the data types are coorect.

In [181]:
# print the information of `world_happiness_2016` dataframe
world_happiness_2016.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        157 non-null    object 
 1   Region                         157 non-null    object 
 2   Happiness Rank                 157 non-null    int64  
 3   Happiness Score                157 non-null    float64
 4   Lower Confidence Interval      157 non-null    float64
 5   Upper Confidence Interval      157 non-null    float64
 6   Economy (GDP per Capita)       157 non-null    float64
 7   Family                         157 non-null    float64
 8   Health (Life Expectancy)       157 non-null    float64
 9   Freedom                        157 non-null    float64
 10  Trust (Government Corruption)  157 non-null    float64
 11  Generosity                     157 non-null    float64
 12  Dystopia Residual              157 non-null    flo

There are no missing values and the data types are coorect.

In [182]:
# print the information of `world_happiness_2017` dataframe
world_happiness_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155 entries, 0 to 154
Data columns (total 12 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        155 non-null    object 
 1   Happiness.Rank                 155 non-null    int64  
 2   Happiness.Score                155 non-null    float64
 3   Whisker.high                   155 non-null    float64
 4   Whisker.low                    155 non-null    float64
 5   Economy..GDP.per.Capita.       155 non-null    float64
 6   Family                         155 non-null    float64
 7   Health..Life.Expectancy.       155 non-null    float64
 8   Freedom                        155 non-null    float64
 9   Generosity                     155 non-null    float64
 10  Trust..Government.Corruption.  155 non-null    float64
 11  Dystopia.Residual              155 non-null    float64
dtypes: float64(10), int64(1), object(1)
memory usage: 

There are no missing values and the data types are coorect.

In [183]:
# print the information of `world_happiness_2018` dataframe
world_happiness_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    object 
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     155 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 11.1+ KB


There are no missing values in all columns except `Perceptions of corruption` column only one missing value and the data types are coorect for all.

In [184]:
# print the information of `world_happiness_2019` dataframe
world_happiness_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    object 
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     156 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 11.1+ KB


There are no missing values and the data types are coorect.

In [185]:
# print the descriptive statistics summary for `world_happiness_2015` dataframe
world_happiness_2015.describe(include="object") # print only for columns with object data type

Country              Region
count           158                 158
unique          158                  10
top     Switzerland  Sub-Saharan Africa
freq              1                  40

In [186]:
# print the descriptive statistics summary for `world_happiness_2016` dataframe
world_happiness_2016.describe(include="object")

Country              Region
count       157                 157
unique      157                  10
top     Denmark  Sub-Saharan Africa
freq          1                  38

In [187]:
# print the descriptive statistics summary for `world_happiness_2017` dataframe
world_happiness_2017.describe(include="object")

Country
count      155
unique     155
top     Norway
freq         1

In [188]:
# print the descriptive statistics summary for `world_happiness_2018` dataframe
world_happiness_2018.describe(include="object")

Country or region
count                156
unique               156
top              Finland
freq                   1

In [189]:
# print the descriptive statistics summary for `world_happiness_2019` dataframe
world_happiness_2019.describe(include="object")

Country or region
count                156
unique               156
top              Finland
freq                   1

In [190]:
# print the descriptive statistics summary for `world_happiness_2015` dataframe
world_happiness_2015.describe(include="number") # print only for columns with numirical data type

Happiness Rank  Happiness Score  Standard Error  \
count      158.000000       158.000000      158.000000   
mean        79.493671         5.375734        0.047885   
std         45.754363         1.145010        0.017146   
min          1.000000         2.839000        0.018480   
25%         40.250000         4.526000        0.037268   
50%         79.500000         5.232500        0.043940   
75%        118.750000         6.243750        0.052300   
max        158.000000         7.587000        0.136930   

       Economy (GDP per Capita)      Family  Health (Life Expectancy)  \
count                158.000000  158.000000                158.000000   
mean                   0.846137    0.991046                  0.630259   
std                    0.403121    0.272369                  0.247078   
min                    0.000000    0.000000                  0.000000   
25%                    0.545808    0.856823                  0.439185   
50%                    0.910245    1.029510                  0.696705   
75%                    1.158448    1.214405                  0.811013   
max                    1.690420    1.402230                  1.025250   

          Freedom  Trust (Government Corruption)  Generosity  \
count  158.000000                     158.000000  158.000000   
mean     0.428615                       0.143422    0.237296   
std      0.150693                       0.120034    0.126685   
min      0.000000                       0.000000    0.000000   
25%      0.328330                       0.061675    0.150553   
50%      0.435515                       0.107220    0.216130   
75%      0.549092                       0.180255    0.309883   
max      0.669730                       0.551910    0.795880   

       Dystopia Residual  
count         158.000000  
mean            2.098977  
std             0.553550  
min             0.328580  
25%             1.759410  
50%             2.095415  
75%             2.462415  
max             3.602140

In [191]:
# print the descriptive statistics summary for `world_happiness_2016` dataframe
world_happiness_2016.describe(include="number")

Happiness Rank  Happiness Score  Lower Confidence Interval  \
count      157.000000       157.000000                 157.000000   
mean        78.980892         5.382185                   5.282395   
std         45.466030         1.141674                   1.148043   
min          1.000000         2.905000                   2.732000   
25%         40.000000         4.404000                   4.327000   
50%         79.000000         5.314000                   5.237000   
75%        118.000000         6.269000                   6.154000   
max        157.000000         7.526000                   7.460000   

       Upper Confidence Interval  Economy (GDP per Capita)      Family  \
count                 157.000000                157.000000  157.000000   
mean                    5.481975                  0.953880    0.793621   
std                     1.136493                  0.412595    0.266706   
min                     3.078000                  0.000000    0.000000   
25%                     4.465000                  0.670240    0.641840   
50%                     5.419000                  1.027800    0.841420   
75%                     6.434000                  1.279640    1.021520   
max                     7.669000                  1.824270    1.183260   

       Health (Life Expectancy)     Freedom  Trust (Government Corruption)  \
count                157.000000  157.000000                     157.000000   
mean                   0.557619    0.370994                       0.137624   
std                    0.229349    0.145507                       0.111038   
min                    0.000000    0.000000                       0.000000   
25%                    0.382910    0.257480                       0.061260   
50%                    0.596590    0.397470                       0.105470   
75%                    0.729930    0.484530                       0.175540   
max                    0.952770    0.608480                       0.505210   

       Generosity  Dystopia Residual  
count  157.000000         157.000000  
mean     0.242635           2.325807  
std      0.133756           0.542220  
min      0.000000           0.817890  
25%      0.154570           2.031710  
50%      0.222450           2.290740  
75%      0.311850           2.664650  
max      0.819710           3.837720

In [192]:
# print the descriptive statistics summary for `world_happiness_2017` dataframe
world_happiness_2017.describe(include="number")

Happiness.Rank  Happiness.Score  Whisker.high  Whisker.low  \
count      155.000000       155.000000    155.000000   155.000000   
mean        78.000000         5.354019      5.452326     5.255713   
std         44.888751         1.131230      1.118542     1.145030   
min          1.000000         2.693000      2.864884     2.521116   
25%         39.500000         4.505500      4.608172     4.374955   
50%         78.000000         5.279000      5.370032     5.193152   
75%        116.500000         6.101500      6.194600     6.006527   
max        155.000000         7.537000      7.622030     7.479556   

       Economy..GDP.per.Capita.      Family  Health..Life.Expectancy.  \
count                155.000000  155.000000                155.000000   
mean                   0.984718    1.188898                  0.551341   
std                    0.420793    0.287263                  0.237073   
min                    0.000000    0.000000                  0.000000   
25%                    0.663371    1.042635                  0.369866   
50%                    1.064578    1.253918                  0.606042   
75%                    1.318027    1.414316                  0.723008   
max                    1.870766    1.610574                  0.949492   

          Freedom  Generosity  Trust..Government.Corruption.  \
count  155.000000  155.000000                     155.000000   
mean     0.408786    0.246883                       0.123120   
std      0.149997    0.134780                       0.101661   
min      0.000000    0.000000                       0.000000   
25%      0.303677    0.154106                       0.057271   
50%      0.437454    0.231538                       0.089848   
75%      0.516561    0.323762                       0.153296   
max      0.658249    0.838075                       0.464308   

       Dystopia.Residual  
count         155.000000  
mean            1.850238  
std             0.500028  
min             0.377914  
25%             1.591291  
50%             1.832910  
75%             2.144654  
max             3.117485

In [193]:
# print the descriptive statistics summary for `world_happiness_2018` dataframe
world_happiness_2018.describe(include="number")

Overall rank       Score  GDP per capita  Social support  \
count    156.000000  156.000000      156.000000      156.000000   
mean      78.500000    5.375917        0.891449        1.213237   
std       45.177428    1.119506        0.391921        0.302372   
min        1.000000    2.905000        0.000000        0.000000   
25%       39.750000    4.453750        0.616250        1.066750   
50%       78.500000    5.378000        0.949500        1.255000   
75%      117.250000    6.168500        1.197750        1.463000   
max      156.000000    7.632000        2.096000        1.644000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               156.000000                    156.000000  156.000000   
mean                  0.597346                      0.454506    0.181006   
std                   0.247579                      0.162424    0.098471   
min                   0.000000                      0.000000    0.000000   
25%                   0.422250                      0.356000    0.109500   
50%                   0.644000                      0.487000    0.174000   
75%                   0.777250                      0.578500    0.239000   
max                   1.030000                      0.724000    0.598000   

       Perceptions of corruption  
count                 155.000000  
mean                    0.112000  
std                     0.096492  
min                     0.000000  
25%                     0.051000  
50%                     0.082000  
75%                     0.137000  
max                     0.457000

In [194]:
# print the descriptive statistics summary for `world_happiness_2019` dataframe
world_happiness_2019.describe(include="number")

Overall rank       Score  GDP per capita  Social support  \
count    156.000000  156.000000      156.000000      156.000000   
mean      78.500000    5.407096        0.905147        1.208814   
std       45.177428    1.113120        0.398389        0.299191   
min        1.000000    2.853000        0.000000        0.000000   
25%       39.750000    4.544500        0.602750        1.055750   
50%       78.500000    5.379500        0.960000        1.271500   
75%      117.250000    6.184500        1.232500        1.452500   
max      156.000000    7.769000        1.684000        1.624000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               156.000000                    156.000000  156.000000   
mean                  0.725244                      0.392571    0.184846   
std                   0.242124                      0.143289    0.095254   
min                   0.000000                      0.000000    0.000000   
25%                   0.547750                      0.308000    0.108750   
50%                   0.789000                      0.417000    0.177500   
75%                   0.881750                      0.507250    0.248250   
max                   1.141000                      0.631000    0.566000   

       Perceptions of corruption  
count                 156.000000  
mean                    0.110603  
std                     0.094538  
min                     0.000000  
25%                     0.047000  
50%                     0.085500  
75%                     0.141250  
max                     0.453000

The process of profiling differs slightly for categorical and numerical variables due to their inherent differences.

**The two main types of data are:**
- Quantitative (numerical) data
- Qualitative (categorical) data

### Data Quality Checks
Data quality checks involve the process of ensuring that the data is accurate, complete, consistent, relevant, and reliable.


**Here are typical steps involved in checking data quality:**

#### 1. Reliability:
Evaluate the data's source and collection process to determine its trustworthiness.

> The data source appears to be reliable as it comes from the [World Happiness Report](https://worldhappiness.report/), which is prepared by [Sustainable Development Solutions Network](https://www.unsdsn.org/) in collaboration with experts from various fields. The report is based on data from the [Gallup World Poll](https://www.gallup.com/analytics/318875/global-research.aspx), a trusted global survey, and uses recognized research methodologies.

#### 2. Timeliness:
Ensure the data is up-to-date and reflective of the current situation or the period of interest for the analysis.

>Because the questions in the problem statement for this usecase require comparing data between specific years (2015, 2016, 2017), and also over the years. So, the data covers that and is relevant to this period and is considered appropriate and suitable for the required time analysis.

#### 3. Consistency:

Confirm that the data is consistent within the dataset and across multiple data sources. For example, the same data point should not have different values in different places.


> There is no consistency issue in the dataset.

#### 4. Relevance:
Assess whether the data is appropriate and applicable for the intended analysis. Data that is not relevant can skew results and lead to incorrect conclusions.

**Key considerations for relevance include:**

> 1. Sample Appropriateness: Confirm that your data sample aligns with your analysis objectives. For instance, utilizing data from the Northern region will not yield accurate insights for the Western region of the Kingdom.



>
> 2. Variable Selection: Any column will not be relevant for our analysis, we can get rid of these using the drop() method. We will set the “axis” argument to 1 since we’re dealing with columns, and set the “inplace” argument to True to make the change permanent.

>


**1- Sample Appropriateness:**

The sample data for all datasets are Appropriate to analysis objectives.


**2- Variable Selection:**

Below, We will check the columns that is useful for our analysis and will help us to answer the problem statement questions.  

In [195]:
world_happiness_2015.head(3)

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938                        0.48357   

   Generosity  Dystopia Residual  
0     0.29678            2.51738  
1     0.43630            2.70201  
2     0.34139            2.49204

In [196]:
world_happiness_2016.head(3)

Country          Region  Happiness Rank  Happiness Score  \
0      Denmark  Western Europe               1            7.526   
1  Switzerland  Western Europe               2            7.509   
2      Iceland  Western Europe               3            7.501   

   Lower Confidence Interval  Upper Confidence Interval  \
0                      7.460                      7.592   
1                      7.428                      7.590   
2                      7.333                      7.669   

   Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
0                   1.44178  1.16374                   0.79504  0.57941   
1                   1.52733  1.14524                   0.86303  0.58557   
2                   1.42666  1.18326                   0.86733  0.56624   

   Trust (Government Corruption)  Generosity  Dystopia Residual  
0                        0.44453     0.36171            2.73939  
1                        0.41203     0.28083            2.69463  
2                        0.14975     0.47678            2.83137

In [197]:
world_happiness_2017.head(3)

Country  Happiness.Rank  Happiness.Score  Whisker.high  Whisker.low  \
0   Norway               1            7.537      7.594445     7.479556   
1  Denmark               2            7.522      7.581728     7.462272   
2  Iceland               3            7.504      7.622030     7.385970   

   Economy..GDP.per.Capita.    Family  Health..Life.Expectancy.   Freedom  \
0                  1.616463  1.533524                  0.796667  0.635423   
1                  1.482383  1.551122                  0.792566  0.626007   
2                  1.480633  1.610574                  0.833552  0.627163   

   Generosity  Trust..Government.Corruption.  Dystopia.Residual  
0    0.362012                       0.315964           2.277027  
1    0.355280                       0.400770           2.313707  
2    0.475540                       0.153527           2.322715

In [198]:
world_happiness_2018.head(3)

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.632           1.305           1.592   
1             2            Norway  7.594           1.456           1.582   
2             3           Denmark  7.555           1.351           1.590   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.874                         0.681       0.202   
1                    0.861                         0.686       0.286   
2                    0.868                         0.683       0.284   

   Perceptions of corruption  
0                      0.393  
1                      0.340  
2                      0.408

In [199]:
world_happiness_2019.head(3)

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.769           1.340           1.587   
1             2           Denmark  7.600           1.383           1.573   
2             3            Norway  7.554           1.488           1.582   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.986                         0.596       0.153   
1                    0.996                         0.592       0.252   
2                    1.028                         0.603       0.271   

   Perceptions of corruption  
0                      0.393  
1                      0.410  
2                      0.341

There are columns unrelated to your project questions, which focus on:
- Ranking countries in happiness.
- The six factors contributing to happiness (GDP per capita, social support, healthy life expectancy, freedom, generosity, and perceptions of corruption).
- Changes in country rankings over the years.

**The columns to drop are::**
- **2015 and 2016:**
  - `Standard Error` (2015)
  - `Lower Confidence Interval` and `Upper Confidence Interval` (2016)
- **2017:**
  - `Whisker.high` and `Whisker.low` (statistical bounds, not core happiness factors)
- **2018 and 2019:**
  - All columns are core factors; no drops needed.



#### 5. Uniqueness:
Check for and remove duplicate records to prevent skewed analysis results.


In [200]:
world_happiness_2017[world_happiness_2017.duplicated(['Happiness.Rank'], keep=False)]

Empty DataFrame
Columns: [Country, Happiness.Rank, Happiness.Score, Whisker.high, Whisker.low, Economy..GDP.per.Capita., Family, Health..Life.Expectancy., Freedom, Generosity, Trust..Government.Corruption., Dystopia.Residual]
Index: []

In [201]:
world_happiness_2017[world_happiness_2017.duplicated(['Country'], keep=False)]

Empty DataFrame
Columns: [Country, Happiness.Rank, Happiness.Score, Whisker.high, Whisker.low, Economy..GDP.per.Capita., Family, Health..Life.Expectancy., Freedom, Generosity, Trust..Government.Corruption., Dystopia.Residual]
Index: []

In [202]:
world_happiness_2015.duplicated().sum()

0

In [203]:
world_happiness_2016.duplicated().sum()

0

In [204]:
world_happiness_2017.duplicated().sum()

0

In [205]:
world_happiness_2018.duplicated().sum()

0

In [206]:
world_happiness_2019.duplicated().sum()

0

In [207]:
world_happiness_2015.duplicated(["Happiness Rank"]).sum()

1

In [208]:
world_happiness_2015[world_happiness_2015.duplicated(['Happiness Rank'], keep=False)]

Country                           Region  Happiness Rank  \
81      Jordan  Middle East and Northern Africa              82   
82  Montenegro       Central and Eastern Europe              82   

    Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
81            5.192         0.04524                   0.90198  1.05392   
82            5.192         0.05235                   0.97438  0.90557   

    Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
81                   0.69639  0.40661                        0.14293   
82                   0.72521  0.18260                        0.14296   

    Generosity  Dystopia Residual  
81     0.11053            1.87996  
82     0.16140            2.10017

In [209]:
world_happiness_2016.duplicated(["Happiness Rank"]).sum()

3

In [210]:
world_happiness_2016[world_happiness_2016.duplicated(['Happiness Rank'], keep=False)]

Country                           Region  Happiness Rank  \
33   Saudi Arabia  Middle East and Northern Africa              34   
34         Taiwan                     Eastern Asia              34   
56         Poland       Central and Eastern Europe              57   
57    South Korea                     Eastern Asia              57   
144  Burkina Faso               Sub-Saharan Africa             145   
145        Uganda               Sub-Saharan Africa             145   

     Happiness Score  Lower Confidence Interval  Upper Confidence Interval  \
33             6.379                      6.287                      6.471   
34             6.379                      6.305                      6.453   
56             5.835                      5.749                      5.921   
57             5.835                      5.747                      5.923   
144            3.739                      3.647                      3.831   
145            3.739                      3.629                      3.849   

     Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
33                    1.48953  0.84829                   0.59267  0.37904   
34                    1.39729  0.92624                   0.79565  0.32377   
56                    1.24585  1.04685                   0.69058  0.45190   
57                    1.35948  0.72194                   0.88645  0.25168   
144                   0.31995  0.63054                   0.21297  0.33370   
145                   0.34719  0.90981                   0.19625  0.43653   

     Trust (Government Corruption)  Generosity  Dystopia Residual  
33                         0.30008     0.15457            2.61482  
34                         0.06630     0.25495            2.61523  
56                         0.05500     0.14443            2.20035  
57                         0.07716     0.18824            2.35015  
144                        0.12533     0.24353            1.87319  
145                        0.06442     0.27102            1.51416

In [211]:
world_happiness_2017.duplicated(["Happiness.Rank"]).sum()

0

In [212]:
world_happiness_2018.duplicated(["Overall rank"]).sum()

0

In [213]:
world_happiness_2019.duplicated(["Overall rank"]).sum()

0

- **No duplicate rows** exist in any of the datasets when checking all columns (`duplicated().sum() == 0` for all years).
- **Checking for duplicates in ranking columns:**
  - **2015:** 1 duplicate in `Happiness Rank`.
  - **2016:** 3 duplicates in `Happiness Rank`.
  - **2017, 2018, 2019:** No duplicates in `Happiness Rank` or `Overall rank`.

These duplicates are due to tied rankings rather than actual duplicate rows.
**2015:**
- **Rank 82 is duplicated between:**
  - **Jordan** and **Montenegro**.
- Despite having the same rank, they differ in values such as social support, health, and freedom, indicating a tie in ranking rather than duplicate data.

**2016:**
- **Rank 34 is duplicated between:**
  - **Saudi Arabia** and **Taiwan**.
- **Rank 57 is duplicated between:**
  - **Poland** and **South Korea**.
- **Rank 145 is duplicated between:**
  - **Burkina Faso** and **Uganda**.
- All these countries share the same rank but have different values for other factors.

So, We do not delete these rows as they are not duplicates but tied rankings.
- Retain them as they accurately reflect equal happiness rankings among different countries, which is a significant aspect of the analysis.



#### 6. Completeness:
Ensure that no critical data is missing. This might mean checking for null values or required fields that are empty.

We will start by checking the dataset for missing or null values. For this, we can use the isna() method which returns a dataframe of boolean values indicating if a field is null or not. To group all missing values by column, we can include the sum() method.

In [214]:
print("Missing values in world_happiness_2015:", world_happiness_2015.isnull().sum().sum())
print("Missing values in world_happiness_2016:", world_happiness_2016.isnull().sum().sum())
print("Missing values in world_happiness_2017:", world_happiness_2017.isnull().sum().sum())
print("Missing values in world_happiness_2018:", world_happiness_2018.isnull().sum().sum())
print("Missing values in world_happiness_2019:", world_happiness_2019.isnull().sum().sum())

Missing values in world_happiness_2015: 0
Missing values in world_happiness_2016: 0
Missing values in world_happiness_2017: 0
Missing values in world_happiness_2018: 1
Missing values in world_happiness_2019: 0


In [215]:
# Count total occurrences of '-' in each dataset:
datasets = [world_happiness_2015, world_happiness_2016, world_happiness_2017,
            world_happiness_2018, world_happiness_2019]

dataset_names = ["world_happiness_2015", "world_happiness_2016", "world_happiness_2017",
                 "world_happiness_2018", "world_happiness_2019"]

for name, df in zip(dataset_names, datasets):
    print(f"Occurrences of '-' in {name}:")
    print((df == '-').sum())
    print("-" * 40)

for name, df in zip(dataset_names, datasets):
    print(f"Total occurrences of '-' in {name}: {df.isin(['-']).sum().sum()}")

Occurrences of '-' in world_happiness_2015:
Country                          0
Region                           0
Happiness Rank                   0
Happiness Score                  0
Standard Error                   0
Economy (GDP per Capita)         0
Family                           0
Health (Life Expectancy)         0
Freedom                          0
Trust (Government Corruption)    0
Generosity                       0
Dystopia Residual                0
dtype: int64
----------------------------------------
Occurrences of '-' in world_happiness_2016:
Country                          0
Region                           0
Happiness Rank                   0
Happiness Score                  0
Lower Confidence Interval        0
Upper Confidence Interval        0
Economy (GDP per Capita)         0
Family                           0
Health (Life Expectancy)         0
Freedom                          0
Trust (Government Corruption)    0
Generosity                       0
Dystopia Residual 

There is only one missing value, and removing it will not significantly affect the analysis, especially since it is spread across hundreds of rows. Therefore, we will delete it.

#### 7. Check Accuracy:

Verify that the data is correct and precise. This could involve comparing data samples with known sources or using validation rules.

**The process includes:**
1. Validating the appropriateness of data types for the dataset.
2. Identifying outliers  using established validation  rule

In [216]:
# check columns types
world_happiness_2015.dtypes

Country                           object
Region                            object
Happiness Rank                     int64
Happiness Score                  float64
Standard Error                   float64
Economy (GDP per Capita)         float64
Family                           float64
Health (Life Expectancy)         float64
Freedom                          float64
Trust (Government Corruption)    float64
Generosity                       float64
Dystopia Residual                float64
dtype: object

In [217]:
# check columns types
world_happiness_2016.dtypes

Country                           object
Region                            object
Happiness Rank                     int64
Happiness Score                  float64
Lower Confidence Interval        float64
Upper Confidence Interval        float64
Economy (GDP per Capita)         float64
Family                           float64
Health (Life Expectancy)         float64
Freedom                          float64
Trust (Government Corruption)    float64
Generosity                       float64
Dystopia Residual                float64
dtype: object

In [218]:
# check columns types
world_happiness_2017.dtypes

Country                           object
Happiness.Rank                     int64
Happiness.Score                  float64
Whisker.high                     float64
Whisker.low                      float64
Economy..GDP.per.Capita.         float64
Family                           float64
Health..Life.Expectancy.         float64
Freedom                          float64
Generosity                       float64
Trust..Government.Corruption.    float64
Dystopia.Residual                float64
dtype: object

In [219]:
# check columns types
world_happiness_2018.dtypes

Overall rank                      int64
Country or region                object
Score                           float64
GDP per capita                  float64
Social support                  float64
Healthy life expectancy         float64
Freedom to make life choices    float64
Generosity                      float64
Perceptions of corruption       float64
dtype: object

In [220]:
# check columns types
world_happiness_2019.dtypes

Overall rank                      int64
Country or region                object
Score                           float64
GDP per capita                  float64
Social support                  float64
Healthy life expectancy         float64
Freedom to make life choices    float64
Generosity                      float64
Perceptions of corruption       float64
dtype: object

No format outliers.

In [221]:
# check outliers

### **Detect Outliers Using IQR:**

# Function to detect outliers using the IQR method
def detect_outliers(df, columns):
    outliers = {}
    for col in columns:
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR
        outliers[col] = df[(df[col] < lower_bound) | (df[col] > upper_bound)]
    return outliers

# Select numeric columns for outlier detection
numeric_columns = {
    "2015": world_happiness_2015.select_dtypes(include=[np.number]).columns,
    "2016": world_happiness_2016.select_dtypes(include=[np.number]).columns,
    "2017": world_happiness_2017.select_dtypes(include=[np.number]).columns,
    "2018": world_happiness_2018.select_dtypes(include=[np.number]).columns,
    "2019": world_happiness_2019.select_dtypes(include=[np.number]).columns
}

# Detect outliers for each dataset
outliers_detected = {
    "2015": detect_outliers(world_happiness_2015, numeric_columns["2015"]),
    "2016": detect_outliers(world_happiness_2016, numeric_columns["2016"]),
    "2017": detect_outliers(world_happiness_2017, numeric_columns["2017"]),
    "2018": detect_outliers(world_happiness_2018, numeric_columns["2018"]),
    "2019": detect_outliers(world_happiness_2019, numeric_columns["2019"])
}

# Count the number of outliers per column per dataset
outliers_summary = {
    year: {col: len(outliers[col]) for col in outliers}
    for year, outliers in outliers_detected.items()
}

# Display the summary of outliers detected
import pandas as pd
outliers_summary_df = pd.DataFrame(outliers_summary)
print(outliers_summary_df)


                               2015  2016  2017  2018  2019
Happiness Rank                  0.0   0.0   NaN   NaN   NaN
Happiness Score                 0.0   0.0   NaN   NaN   NaN
Standard Error                 10.0   NaN   NaN   NaN   NaN
Economy (GDP per Capita)        0.0   0.0   NaN   NaN   NaN
Family                          3.0   1.0   4.0   NaN   NaN
Health (Life Expectancy)        0.0   0.0   NaN   NaN   NaN
Freedom                         0.0   0.0   0.0   NaN   NaN
Trust (Government Corruption)  14.0  12.0   NaN   NaN   NaN
Generosity                      2.0   4.0   2.0   2.0   2.0
Dystopia Residual               5.0   6.0   NaN   NaN   NaN
Lower Confidence Interval       NaN   0.0   NaN   NaN   NaN
Upper Confidence Interval       NaN   0.0   NaN   NaN   NaN
Happiness.Rank                  NaN   NaN   0.0   NaN   NaN
Happiness.Score                 NaN   NaN   0.0   NaN   NaN
Whisker.high                    NaN   NaN   0.0   NaN   NaN
Whisker.low                     NaN   Na

In [222]:
# go to clean them

**What is an Outlier?**
Outlier is an row/observation that appears far away and diverges from an overall pattern in a sample.

**What are the types of Outliers?**
1. Univariate: These outliers can be found when we look at distribution of a single variable
2. Multivariate: are outliers in an n-dimensional space. In order to find them, you have to look at distributions in multi-dimensions. example (hight=100, weight=100) for a person

**What causes Outliers?**
Whenever we come across outliers, the ideal way to tackle them is to find out the reason of having these outliers. The method to deal with them would then depend on the reason of their occurrence.

Let’s understand various types of outliers:

1. Data Entry Errors:- Human errors such as errors caused during data collection, recording, or entry can cause outliers in data.
2. Measurement Error: It is the most common source of outliers. This is caused when the measurement instrument used turns out to be faulty.
3. Data Processing Error: Whenever we perform data mining, we extract data from multiple sources. It is possible that some manipulation or extraction errors may lead to outliers in the dataset.
4. Sampling error: For instance, we have to measure the height of athletes. By mistake, we include a few basketball players in the sample. This inclusion is likely to cause outliers in the dataset.
5. Natural Outlier: When an outlier is not artificial (due to error), it is a natural outlier. For instance: In my last assignment with one of the renowned insurance company, I noticed that the performance of top 50 financial advisors was far higher than rest of the population. Surprisingly, it was not due to any error. Hence, whenever we perform any data mining activity with advisors, we used to treat this segment separately.


**What is the impact of Outliers on a dataset?**


![image.png](https://www.analyticsvidhya.com/wp-content/uploads/2015/02/Outlier_31.png)



**How to detect Outliers?**

1. Most commonly used method to detect outliers is visualization (Univariate Graphical Analysis).

We use 3 common visualization methods:
>- Box-plot: A box plot is a method for graphically depicting groups of numerical data through their quartiles. The box extends from the Q1 to Q3 quartile values of the data, with a line at the median (Q2). The whiskers extend from the edges of the box to show the range of the data. Outlier points are those past the end of the whiskers. Box plots show robust measures of location and spread as well as providing information about symmetry and outliers.
>
>  
>![image.png](https://miro.medium.com/v2/resize:fit:698/format:webp/1*VK5iHA2AB28HSZwWwUbNYg.png)
>
>
>- Histogram
>- Scatter Plot: A scatter plot is a mathematical diagram using Cartesian coordinates to display values for two variables for a set of data. The data are displayed as a collection of points, each having the value of one variable determining the position on the horizontal axis and the value of the other variable determining the position on the vertical axis. The points that are far from the population can be termed as an outlier.
>
>  
>![image.png](https://miro.medium.com/v2/resize:fit:4800/format:webp/1*Ov6aH-8yIwNoUxtMFwgx4g.png)
>
>

2. Using statistical method (Univariate Non-Graphical analysis):
>- Any value, which is beyond the range of -1.5 x IQR to 1.5 x IQR

![image.png](https://www.whatissixsigma.net/wp-content/uploads/2015/07/Box-Plot-Diagram-to-identify-Outliers-figure-1.png)

>- Use capping methods. Any value which out of range of 5th and 95th percentile can be considered as outlier
>- Data points, three or more standard deviation away from mean are considered outlier: The Z-score is the signed number of standard deviations by which the value of an observation or data point is above the mean value of what is being observed or measured. While calculating the Z-score we re-scale and center the data and look for data points that are too far from zero. These data points which are way too far from zero will be treated as the outliers. In most of the cases, a threshold of 3 or -3 is used i.e if the Z-score value is greater than or less than 3 or -3 respectively, that data point will be identified as outliers.
> - Outlier detection is merely a special case of the examination of data for influential data points and it also depends on the business understanding


In [223]:
# go to univariate graphical analysis
# go to lesson : data visualisation 1 - chart type section
# then go to univariate graphical analysis
# detect outliers using graphs varbaly

In [224]:
# go to lesson: statistics 1 then statistics 3
# then go to univariate Non graphical analysis
# detect outliers using numerical statistics

In [225]:
# go to delete ouliers

## 2. Data Cleaning:

Preliminary findings from data profiling can lead to cleaning the data by:
- Handling missing values
- Correcting errors.
- Dealing with outliers.

-------------------



### Handling missing values:

**Why my data has missing values?**
They may occur at two stages:
1. Data Extraction: It is possible that there are problems with extraction process. Errors at data extraction stage are typically easy to find and can be corrected easily as well.
2. Data collection: These errors occur at time of data collection and are harder to correct.

**Why do we need to handle the missing data?**
To avoid:
- Bias the conclusions.
- Leading the business to make wrong decisions.

**Which are the methods to treat missing values ?**
1. Deletion: we delete rows where any of the variable is missing. Simplicity is one of the major advantage of this method, but this method reduces the power of model because it reduces the sample size.

2. Imputation: is a method to fill in the missing values with estimated ones. This imputation is one of the most frequently used methods.

    2.1. Mean/ Mode/ Median Imputation: It consists of replacing the missing data for a given attribute by the mean or median (quantitative attribute) or mode (qualitative attribute) of all known values of that variable.
    > It can be of two types:
    > - Generalized Imputation: In this case, we calculate the mean or median for all non missing values of that variable then replace missing value with mean or median.
    > - Similar case Imputation: In this case, we calculate average for each group individually of non missing values then replace the missing value based on the group.

    2.2. Constant Value
   
    2.3. Forward Filling
   
    2.4. Backward Filling

6. Prediction Model:  Prediction model is one of the sophisticated method for handling missing data. Here, we create a predictive model to estimate values that will substitute the missing data.  In this case, we divide our data set into two sets: One set with no missing values for the variable and another one with missing values. First data set become training data set of the model while second data set with missing values is test data set and variable with missing values is treated as target variable. Next, we create a model to predict target variable based on other attributes of the training data set and populate missing values of test data set.

> There are 2 drawbacks for this approach:
> - The model estimated values are usually more well-behaved than the true values
> - If there are no relationships with attributes in the data set and the attribute with missing values, then the model will not be precise for estimating missing values.

9. KNN Imputation: In this method of imputation, the missing values of an attribute are imputed using the given number of attributes that are most similar to the attribute whose values are missing. The similarity of two attributes is determined using a distance function. It is also known to have certain advantage & disadvantages.

   > **Advantages:**
   > - k-nearest neighbour can predict both qualitative & quantitative attributes
   > - Creation of predictive model for each attribute with missing data is not required
   > - Attributes with multiple missing values can be easily treated
   > - Correlation structure of the data is taken into consideration

   > **Disadvantage:**
   > - KNN algorithm is very time-consuming in analyzing large database. It searches through all the dataset looking for the most similar instances.
   > - Choice of k-value is very critical. Higher value of k would include attributes which are significantly different from what we need whereas lower value of k implies missing out of significant attributes.

--------------------


In [226]:
# Standardize column names across datasets
columns_rename = {
    "2015": {
        "Country": "Country",
        "Region": "Region",
        "Happiness Rank": "Rank",
        "Happiness Score": "Score",
        "Economy (GDP per Capita)": "GDP per capita",
        "Family": "Social support",
        "Health (Life Expectancy)": "Healthy life expectancy",
        "Freedom": "Freedom to make life choices",
        "Trust (Government Corruption)": "Perceptions of corruption",
        "Generosity": "Generosity"
    },
    "2016": {
        "Country": "Country",
        "Region": "Region",
        "Happiness Rank": "Rank",
        "Happiness Score": "Score",
        "Economy (GDP per Capita)": "GDP per capita",
        "Family": "Social support",
        "Health (Life Expectancy)": "Healthy life expectancy",
        "Freedom": "Freedom to make life choices",
        "Trust (Government Corruption)": "Perceptions of corruption",
        "Generosity": "Generosity"
    },
    "2017": {
        "Country": "Country",
        "Happiness.Rank": "Rank",
        "Happiness.Score": "Score",
        "Economy..GDP.per.Capita.": "GDP per capita",
        "Family": "Social support",
        "Health..Life.Expectancy.": "Healthy life expectancy",
        "Freedom": "Freedom to make life choices",
        "Trust..Government.Corruption.": "Perceptions of corruption",
        "Generosity": "Generosity"
    },
    "2018": {
        "Country or region": "Country",
        "Overall rank": "Rank",
        "Score": "Score",
        "GDP per capita": "GDP per capita",
        "Social support": "Social support",
        "Healthy life expectancy": "Healthy life expectancy",
        "Freedom to make life choices": "Freedom to make life choices",
        "Perceptions of corruption": "Perceptions of corruption",
        "Generosity": "Generosity"
    },
    "2019": {
        "Country or region": "Country",
        "Overall rank": "Rank",
        "Score": "Score",
        "GDP per capita": "GDP per capita",
        "Social support": "Social support",
        "Healthy life expectancy": "Healthy life expectancy",
        "Freedom to make life choices": "Freedom to make life choices",
        "Perceptions of corruption": "Perceptions of corruption",
        "Generosity": "Generosity"
    }
}

# Rename columns and keep only relevant ones
for year in data:
    data[year] = data[year].rename(columns=columns_rename[year])
    data[year] = data[year][list(columns_rename[year].values())]


In [227]:
# Dimention --> Relevance

columns_to_drop = {
    "2015": ["Standard Error"],
    "2016": ["Lower Confidence Interval", "Upper Confidence Interval"],
    "2017": ["Whisker.high", "Whisker.low"]
}

world_happiness_2015 = world_happiness_2015.drop(columns=columns_to_drop["2015"])
world_happiness_2016 = world_happiness_2016.drop(columns=columns_to_drop["2016"])
world_happiness_2017 = world_happiness_2017.drop(columns=columns_to_drop["2017"])

In [228]:
# go back to 6th dimention --> Completeness

In [229]:
# Dimention --> Completeness
world_happiness_2018 = world_happiness_2018.dropna()
world_happiness_2018.info()

<class 'pandas.core.frame.DataFrame'>
Index: 155 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  155 non-null    int64  
 1   Country or region             155 non-null    object 
 2   Score                         155 non-null    float64
 3   GDP per capita                155 non-null    float64
 4   Social support                155 non-null    float64
 5   Healthy life expectancy       155 non-null    float64
 6   Freedom to make life choices  155 non-null    float64
 7   Generosity                    155 non-null    float64
 8   Perceptions of corruption     155 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 12.1+ KB


In [230]:
# go back to 7th dimension Accuracy

### Correcting errors

-------------------

### Dealing with outliers:

**How to remove Outliers?**
Most of the ways to deal with outliers are similar to the methods of missing values like deleting rows, transforming them, binning them, treat them as a separate group, imputing values and other statistical methods. Here, we will discuss the common techniques used to deal with outliers:

1. Deleting rows: We delete outlier values if it is due to data entry error, data processing error or outlier rows are very small in numbers. We can also use trimming at both ends to remove outliers.

2. Imputing: Like imputation of missing values, we can also impute outliers. We can use mean, median, mode imputation methods. Before imputing values, we should analyse if it is natural outlier or artificial. If it is artificial, we can go with imputing values. We can also use statistical model to predict values of outlier rows and after that we can impute it with predicted values.

3. Treat separately: If there are significant number of outliers, we should treat them separately in the statistical model. One of the approach is to treat both groups as two different groups and build individual model for both groups and then combine the output.


In [231]:
import numpy as np

datasets = {
    "2015": world_happiness_2015,
    "2016": world_happiness_2016,
    "2017": world_happiness_2017,
    "2018": world_happiness_2018,
    "2019": world_happiness_2019
}

for year, df in datasets.items():
    print(f" Outliers Analysis for {year}")
    for col in df.select_dtypes(include=[np.number]).columns:
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR
        num_outliers = len(df[(df[col] < lower_bound) | (df[col] > upper_bound)])
        total_rows = len(df)
        percentage = (num_outliers / total_rows) * 100
        print(f"  Column: {col} | Outliers: {num_outliers} | Percentage: {percentage:.2f}%")
    print("-" * 50)


 Outliers Analysis for 2015
  Column: Happiness Rank | Outliers: 0 | Percentage: 0.00%
  Column: Happiness Score | Outliers: 0 | Percentage: 0.00%
  Column: Economy (GDP per Capita) | Outliers: 0 | Percentage: 0.00%
  Column: Family | Outliers: 3 | Percentage: 1.90%
  Column: Health (Life Expectancy) | Outliers: 0 | Percentage: 0.00%
  Column: Freedom | Outliers: 0 | Percentage: 0.00%
  Column: Trust (Government Corruption) | Outliers: 14 | Percentage: 8.86%
  Column: Generosity | Outliers: 2 | Percentage: 1.27%
  Column: Dystopia Residual | Outliers: 5 | Percentage: 3.16%
--------------------------------------------------
 Outliers Analysis for 2016
  Column: Happiness Rank | Outliers: 0 | Percentage: 0.00%
  Column: Happiness Score | Outliers: 0 | Percentage: 0.00%
  Column: Economy (GDP per Capita) | Outliers: 0 | Percentage: 0.00%
  Column: Family | Outliers: 1 | Percentage: 0.64%
  Column: Health (Life Expectancy) | Outliers: 0 | Percentage: 0.00%
  Column: Freedom | Outliers: 0 |

**Outlier Analysis Decision Based on Percentage per Column**

**1- Columns where outliers can be removed:**
If the percentage of outliers is **less than 5%**, removing them will not significantly impact the analysis.  
The columns that meet this criterion include:

- **Family** (1.90% - 2.58%)
- **Generosity** (1.27% - 2.55%)
- **Dystopia Residual** (3.16% - 3.87%)
- **Social Support** (2.58%)
- **Freedom to make life choices** (1.94% - 0.64%)
- **Healthy life expectancy** (0.64%)

---

**2- Columns where outliers should NOT be directly removed:**
If the percentage of outliers is **greater than 5%**, removing them may significantly impact the analysis.  
The affected columns are:

- **Trust (Government Corruption)** (7.64% - 8.86%)
- **Perceptions of corruption** (8.97% - 10.97%)

---
**So, We will:**

**1- Remove outliers** from columns where the percentage of outliers is **less than 5%**.  
**2- For columns with a high percentage of outliers (>5%)**, it is better to **replace them with the nearest boundary values (Capping) instead of removing them**.


In [232]:
for year, df in datasets.items():
    print(f"Columns for {year}:")
    print(df.columns)
    print("-" * 50)


Columns for 2015:
Index(['Country', 'Region', 'Happiness Rank', 'Happiness Score',
       'Economy (GDP per Capita)', 'Family', 'Health (Life Expectancy)',
       'Freedom', 'Trust (Government Corruption)', 'Generosity',
       'Dystopia Residual'],
      dtype='object')
--------------------------------------------------
Columns for 2016:
Index(['Country', 'Region', 'Happiness Rank', 'Happiness Score',
       'Economy (GDP per Capita)', 'Family', 'Health (Life Expectancy)',
       'Freedom', 'Trust (Government Corruption)', 'Generosity',
       'Dystopia Residual'],
      dtype='object')
--------------------------------------------------
Columns for 2017:
Index(['Country', 'Happiness.Rank', 'Happiness.Score',
       'Economy..GDP.per.Capita.', 'Family', 'Health..Life.Expectancy.',
       'Freedom', 'Generosity', 'Trust..Government.Corruption.',
       'Dystopia.Residual'],
      dtype='object')
--------------------------------------------------
Columns for 2018:
Index(['Overall rank', 

In [233]:
columns_to_drop_outliers = {
    "2015": ["Family", "Generosity", "Dystopia Residual", "Freedom", "Health (Life Expectancy)"],
    "2016": ["Family", "Generosity", "Dystopia Residual", "Freedom", "Health (Life Expectancy)"],
    "2017": ["Family", "Generosity", "Dystopia.Residual", "Freedom", "Health..Life.Expectancy."],
    "2018": ["Social support", "Generosity", "Dystopia Residual", "Freedom to make life choices", "Healthy life expectancy"],
    "2019": ["Social support", "Generosity", "Dystopia Residual", "Freedom to make life choices", "Healthy life expectancy"]
}

for year, df in datasets.items():
    if year in columns_to_drop_outliers:
        for col in columns_to_drop_outliers[year]:
            if col in df.columns:
                Q1 = df[col].quantile(0.25)
                Q3 = df[col].quantile(0.75)
                IQR = Q3 - Q1
                lower_bound = Q1 - 1.5 * IQR
                upper_bound = Q3 + 1.5 * IQR
                df = df[(df[col] >= lower_bound) & (df[col] <= upper_bound)]
    datasets[year] = df



In [234]:
for year, df in datasets.items():
    if year in columns_to_drop_outliers:
        for col in columns_to_drop_outliers[year]:
            if col in df.columns:
                Q1 = df[col].quantile(0.25)
                Q3 = df[col].quantile(0.75)
                IQR = Q3 - Q1
                lower_bound = Q1 - 1.5 * IQR
                upper_bound = Q3 + 1.5 * IQR
                df = df[(df[col] >= lower_bound) & (df[col] <= upper_bound)]
    datasets[year] = df

## 3. Univariate Analysis:

This involves examining single variables to understand their characteristics (distribution, central tendency, dispersion, and shape).

We calculate **numerical values** about the data that tells us about the distribution of the data. We also **draw graphs** showing visually how the data is distributed. **To answer the following questions about Features/characteristics of Data:**
- Where is the center of the data? (location)
- How much does the data vary? (scale)
- What is the shape of the data? (shape)

**The benefits of this analysis:**
Statistics summary gives a high-level idea to identify whether the data has any outliers, data entry error, distribution of data such as the data is normally distributed or left/right skewed

**In this step, we will explore variables one by one using following approaches:**

### 1. Univariate Graphical Analysis:
Method to perform uni-variate analysis will depend on whether the variable type is categorical or numerical.

#### I. Categorical Variables:

we’ll use frequency table to understand distribution of each category
- Bar Chart (Ordinal) - Orderd
- Pie Chart (Nominal) - non Orderd

#### II. Numerical Variables:

we need to understand the central tendency and spread of the variable (Descriptive Analysis) using:
   - Box plot
   - Histogram

### 2. Univariate Non-Graphical analysis:

- Where is the center of the data? (location) --> **Measures of central tendency**
- How much does the data vary? (scale) --> **Measure of variability**
- What is the shape of the data? (shape) --> **Measures of variation combined with an average (measure of center) gives a good picture of the distribution of the data.**

## 4. Bivariate/Multivariate Analysis:

Here, you look at the relationships between two or more variables. This can involve looking for correlations, patterns, and trends that suggest a relationship or an association.

We can perform bi-variate analysis for any combination of categorical and numerical variables. The combination can be:
| bi-variate variables   | Plot type |
| ------------- | ------------- |
| Categorical & Categorical| Stacked Bar Chart |
| Categorical & numerical  | scatter plot, histogram, box plot|
| numerical  & numerical  | Scatter plot, line chart|


Multivariate Analysis:
- Heat map
- Bar Chart
- Scatter Chart
- Line Chart

**Categorical & Categorical --> (Stacked Column Chart)**

**Categorical & numerical --> (scatter plot, histogram, box plot)**

**numerical & numerical --> (Scatter plot, line chart)**

We could also use a correlation matrix to get more specific information about the relationship between these two variables.

Questions:

1. What countries or regions rank the highest in overall happiness and each of
the six factors contributing to happiness?

2. How did country ranks or scores change between the 2015 and 2016 as well as
the 2016 and 2017 reports?

3. Did any country experience a significant increase or decrease in happiness?

4. Bounce: Please begin your analysis, and don
't hesitate to consider additional
relevant questions.

In [237]:
top_countries = {year: df.nlargest(5, 'Score')[['Country', 'Score']] for year, df in data.items()}
bottom_countries = {year: df.nsmallest(5, 'Score')[['Country', 'Score']] for year, df in data.items()}

rank_changes = {}
for year in range(2015, 2019):
    current_year = data[str(year)]
    next_year = data[str(year + 1)]
    merged = pd.merge(current_year[['Country', 'Rank']], next_year[['Country', 'Rank']], on='Country', suffixes=(f'_{year}', f'_{year+1}'))
    merged['Rank Change'] = merged[f'Rank_{year}'] - merged[f'Rank_{year+1}']
    rank_changes[f'{year}-{year+1}'] = merged[['Country', 'Rank Change']].sort_values(by='Rank Change', ascending=False)

top_countries, bottom_countries, rank_changes


KeyError: '2017'